src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -246,7 +245,7 @@ public RealJenkinsRule withHost(String host) {
                 System.out.println("=== Starting " + description);
                 try {
                     home = tmp.allocate();
-                    port = new Random().nextInt(16384) + 49152; // https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers#Dynamic,_private_or_ephemeral_ports
+                    port = IOUtil.randomTcpPort(); // https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers#Dynamic,_private_or_ephemeral_ports
```

Comment:
Jetty can be start using port 0 which will use any available port as defined by ServerSocket class (https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/net/ServerSocket.html#%3Cinit%3E(int)) then you need the port really used after Jetty has been started. 
see how it's done here https://github.com/jenkinsci/winstone/blob/08016690a4b5d70f97e16769838d7694f13774bc/src/test/java/winstone/HttpConnectorFactoryTest.java#L24
I don't have all the context here but there can be still problem here depending how long it take for the os to really free the port.

Reply:
`RealJenkinsRule` starts Jenkins in a separate jvm, would need a different method to retrieve the port.

Reply:
oh ok. Sounds hackhish to read logs :) 
you should probably the ServerSocket port 0 feature rather than some random.
Look how it's done in Jetty testing https://github.com/eclipse/jetty.project/blob/ec3d7dbe92532cfdbb719cb7b736ad115aea40e5/tests/test-distribution/src/main/java/org/eclipse/jetty/tests/distribution/JettyHomeTester.java#L193

Reply:
Exploring some patch to winstone that would allow to pass a file name where to write the port to.

Reply:
No need for that. You can just have `Endpoint.register` write the port number to a file in `$JENKINS_HOME`. While you are at it, you could have the Jenkins JVM create the random token instead of reading `RealJenkinsRule.token` from a system property, and just write the whole endpoint URL incl. query string to that file: https://github.com/jenkinsci/jenkins-test-harness/blob/959ceb93f5a6cd6927e863de3a5e9b738dd8a051/src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java#L388

**Edit**: no you cannot, because https://github.com/jenkinsci/jenkins-test-harness/blob/959ceb93f5a6cd6927e863de3a5e9b738dd8a051/src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java#L381 would no longer be implementable (and some tests in fact use that method to determine the URL before Jenkins starts).

Reply:
> No need for that.

There is a similar problem for ath, so I'd rather have something that works in both cases.

Reply:
Maybe enhance https://github.com/jenkinsci/jenkins-test-harness/blob/ed2b9ab8b19eaabeb2db699f2db28364c9e96cb0/src/test/java/org/jvnet/hudson/test/RealJenkinsRuleTest.java#L120-L122 to enforce test coverage of the `getUrl` contract.

Reply:
> and some tests in fact use that method to determine the URL before Jenkins starts

Do you have some links?

Reply:
I just wonder whether that contract still makes sense if it prevents foolproof concurrent tests.

Reply:
WIP in https://github.com/jenkinsci/winstone/pull/174 but I can't get incrementals to work...

Reply:
Is there any actual usage of `rr.getUrl()` in tests? I did a quick lookup in `jenkinsci` and couldn't find any beyond the test you linked.

Reply:
There are several usages in @cloudbees that could not easily be rewritten and which I think correspond to a general pattern.

Reply:
ok found one indeed there, seems easy enough for me to rewrite if needed.

Reply:
IIRC I found multiple and did not see a way to rewrite. If you have done a full search, please file PRs removing the `getUrl` call there and CC me and link to this PR. In general I think `getUrl` is useful because you may need to set up WireMock or a Docker fixture of some other service with the Jenkins URL as an initialization parameter (something you cannot modify once the mock service is running).

Reply:
I still think the current patch is safer, but if you want to go this route, I think it would be best to delete `getUrl` (perhaps add a similar method with a different name) as it could no longer be implemented according to its original contract and so any tests previously calling it would need to be adjusted.

Reply:
I guess it could still exist with a different semantic (available as soon as you started Jenkins at least once).

Reply:
> I still think the current patch is safer

I know for a fact that for ATHs this aproach is better but not foolproof.